sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -0,0 +1,406 @@
+                        pkt->observe_value = _decode_uint(pkt_pos, option_len);
+                    }
+                    else {
+                        DEBUG("nanocoap: discarding packet with invalid option length.\n");
+                        return -EBADMSG;
+                    }
+                    break;
+                default:
+                    DEBUG("nanocoap: unhandled option nr=%i len=%i critical=%u\n", option_nr, option_len, option_nr & 1);
+                    if (option_nr & 1) {
+                        DEBUG("nanocoap: discarding packet with unknown critical option.\n");
+                        return -EBADMSG;
+                    }
+            }
+
+            pkt_pos += option_len;
+        }
+    }
+
+    DEBUG("coap pkt parsed. code=%u detail=%u payload_len=%u, 0x%02x\n",
+          coap_get_code_class(pkt),
+          coap_get_code_detail(pkt),
+          pkt->payload_len, hdr->code);
+
+    return 0;
+}
+
+ssize_t coap_handle_req(coap_pkt_t *pkt, uint8_t *resp_buf, unsigned resp_buf_len)
+{
+    if (coap_get_code_class(pkt) != COAP_REQ) {
+        DEBUG("coap_handle_req(): not a request.\n");
+        return -EBADMSG;
+    }
+
+    if (pkt->hdr->code == 0) {
+        return coap_build_reply(pkt, COAP_CODE_EMPTY, resp_buf, resp_buf_len, 0);
+    }
+
+    unsigned method_flag = coap_method2flag(coap_get_code_detail(pkt));
+
+    for (unsigned i = 0; i < coap_resources_numof; i++) {
+        if (!(coap_resources[i].methods & method_flag)) {
+            continue;
+        }
+
+        int res = strcmp((char *)pkt->url, coap_resources[i].path);
+        if (res > 0) {
+            continue;
+        }
+        else if (res < 0) {
+            break;
+        }
+        else {
+            return coap_resources[i].handler(pkt, resp_buf, resp_buf_len);
+        }
+    }
+
+    return coap_build_reply(pkt, COAP_CODE_404, resp_buf, resp_buf_len, 0);
+}
+
+ssize_t coap_reply_simple(coap_pkt_t *pkt,
+                          unsigned code,
+                          uint8_t *buf, size_t len,
+                          unsigned ct,
+                          const uint8_t *payload, uint8_t payload_len)
+{
+    uint8_t *payload_start = buf + coap_get_total_hdr_len(pkt);
+    uint8_t *bufpos = payload_start;
+
+    if (payload_len) {
+        bufpos += coap_put_option_ct(bufpos, 0, ct);
+        *bufpos++ = 0xff;
+
+        memcpy(bufpos, payload, payload_len);
+        bufpos += payload_len;
+    }
+
+    return coap_build_reply(pkt, code, buf, len, bufpos - payload_start);
+}
+
+ssize_t coap_build_reply(coap_pkt_t *pkt, unsigned code,
+                         uint8_t *rbuf, unsigned rlen, unsigned payload_len)
+{
+    unsigned tkl = coap_get_token_len(pkt);
+    unsigned len = sizeof(coap_hdr_t) + tkl;
+
+    if ((len + payload_len + 1) > rlen) {
+        return -ENOSPC;
+    }
+
+    /* if code is COAP_CODE_EMPTY (zero), use RST as type, else RESP */
+    unsigned type = code ? COAP_RESP : COAP_RST;
```

Comment:
What about separate response where we reply with an empty ACK?
This will always turn this into a Reset.